### DISCUSSION TITLE
Correcting data during migration

### DISCUSSION THREAD

**Discussion by monotasker** [MESH-Research/knowledge-commons-repository#80] [General]
There are some obvious errors in the existing CORE data that I'm correcting in the migration script. I'll keep a list here of what I'm doing:

- authors with a CORE role of 'contributor' will be changed to 'contributors' in Invenio *if* there is another author with a more primary role listed
- if the CORE "genre" field is null, a "resource_type" for Invenio will be taken from the CORE "publication-type" field where possible 
- validating ISBN and ISSN numbers and discarding invalid ones (first doing sanitization to strip away extraneous characters)
- lining up different versions of the same user defined keywords (based on normalizing capitalization and junk punctuation) and assigning the set of variations a single id number 
- dropping "chapter" field if it's simply a duplicate of the main "title"
- if "chapter" field is the main "title" preceded by a chapter identifier, keeping just the chapter identifier in the "chapter" field
- stripping surrounding quotation marks and parentheses from "title", "chapter"
- dropping preceding labels like "Chap." for "chapter" field or "Issue no." for "issue" field.
- There are a couple of hundred records with no language. These are *not* all English deposits. I'm using a language detection library (langdetect) to intelligently fill in the language. 
- validating URLs and dropping invalid ones (only a few in practice)

There are some other issues that may need correction, but are judgement calls

- Slashes to escape quotation marks. These end up displaying on the page. Should we replace with html escape characters? 
- Some places we find multiple escape slashes. Should we try to remove? (**At present I'm removing them.**)

**In all cases I'm logging a record of every change, as well as logging instances of uncertain data. We can look at some of the records by hand down the road if we want to.**

**Comment by bjr70** [comment]
Ooh this is interesting. Do you have an example of a deposit where they're all "contributors" with no author listed? I wonder if that's just a language thing? 

What is CORE "genre"? Is that item type? 

>Slashes to escape quotation marks. These end up displaying on the page. Should we replace with html escape characters?

Yes if not too difficult. 

> Some places we find multiple escape slashes. Should we try to remove?

I wonder where they came from? I do think we should try to remove them if we can. If not, that's the way it'll have to be.

**Reply by monotasker** [reply]
CORE has three different fields for item type, often conflicting. "Genre" is the most consistently present and sensible.

Mike agrees about removing the multiple escape slashes. They're consistently junk--artifacts from the process of passing data back and forth between databases, etc.

**Reply by monotasker** [reply]
I don't have an example of the "all contributors" case ready-to-hand. I can dig one up if you like. It didn't seem to be a problem with English, at least in some cases. Looked to me more like they thought "All these people 'contributed'!"

**Reply by bjr70** [reply]
> CORE has three different fields for item type, often conflicting. "Genre" is the most consistently present and sensible.

Ugh. I still don't understand how they made these decisions. 3 fields! o.O [This blog post](https://team.hcommons.org/2021/12/21/five-years-by-the-numbers/) has a list of the top ten item types as of December 2021. I'd say given that list "textDocument" might make sense. Can you run an export of these or can I find them somewhere to look through them? I might be able to assign them.

**Reply by bjr70** [reply]
> I don't have an example of the "all contributors" case ready-to-hand. I can dig one up if you like. It didn't seem to be a problem with English, at least in some cases. Looked to me more like they thought "All these people 'contributed'!"

I wonder if we should just make them "authors"? We're planning on inviting folks to touch up their metadata after launch, so they can adjust as necessary. Anyone else have thoughts? @stephanievaskomsu? @babaklar?

**Reply by monotasker** [reply]
Making them all authors is the way I'm currently going

**Reply by babaklar** [reply]
I tend to think making them all authors is the way to go.

**Reply by bjr70** [reply]
I'll see if Eric can run a query to isolate those deposits to see if we can figure out why they used "contributor" and not author. It is possible they picked "contributor" because the roles listed were much more limited.

**Reply by monotasker** [reply]
> Can you run an export of these or can I find them somewhere to look through them? I might be able to assign them.

Thanks, I've already made tables in the serialization script to line up the CORE values (for each of the 3 fields) with equivalents in our list. So finding equivalents isn't an issue. I just want to flag that I'm using a fallback approach--start with "genre" field, if that's null fall back to "publication-type". But in both cases I'm translating the term into our equivalent.

**Comment by monotasker** [comment]
I edited the top of this discussion with some more of the sanitization changes I'm making. The list isn't complete yet, but for convenience, these are the new ones:

- validating ISBN and ISSN numbers and discarding invalid ones (first doing sanitization to strip away extraneous characters)
- lining up different versions of the same user defined keywords (based on normalizing capitalization and junk punctuation) and assigning the set of variations a single id number
- dropping "chapter" field if it's simply a duplicate of the main "title"
- if "chapter" field is the main "title" preceded by a chapter identifier, keeping just the chapter identifier in the "chapter" field
stripping surrounding quotation marks and parentheses from "title", "chapter" 
- dropping preceding labels like "Chap." for "chapter" field or "Issue no." for "issue" field.

**Comment by rekmla** [comment]
Could you provide an example pid where you do this?

'stripping surrounding quotation marks and parentheses from "title",
"chapter"'


On Fri, May 19, 2023 at 12:26 PM Ian W. Scott ***@***.***>
wrote:

> I edited the top of this discussion with some more of the sanitization
> changes I'm making. The list isn't complete yet, but for convenience, these
> are the new ones:
>
>    - validating ISBN and ISSN numbers and discarding invalid ones (first
>    doing sanitization to strip away extraneous characters)
>    - lining up different versions of the same user defined keywords
>    (based on normalizing capitalization and junk punctuation) and assigning
>    the set of variations a single id number
>    - dropping "chapter" field if it's simply a duplicate of the main
>    "title"
>    - if "chapter" field is the main "title" preceded by a chapter
>    identifier, keeping just the chapter identifier in the "chapter" field
>    stripping surrounding quotation marks and parentheses from "title",
>    "chapter"
>    - dropping preceding labels like "Chap." for "chapter" field or "Issue
>    no." for "issue" field.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MESH-Research/knowledge-commons-repository/discussions/80#discussioncomment-5950519>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB7ZYVUY2TAVIQEEWIMQGPLXG6NLBANCNFSM6AAAAAAX7YLP5E>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID:
> <MESH-Research/knowledge-commons-repository/repo-discussions/80/comments/5950519
> @github.com>
>


-- 
R. Eric Knappe
Head of Software Development
Modern Language Association

www.mla.org

**Reply by monotasker** [reply]
For example, hc:13523 has a chapter field with the value `\"Imposing Pictures: Widow Portraiture as Memorial Strategy in Early Modern Florence\"`

I'm removing quotation marks that frame the entire field, but not removing them when they only contain part of the field. So with `\"The Fiend Gives Friendly Counsel\": Launcelot Gobbo and Polyglot Economics in The Merchant of Venice` I'm leaving the quotation marks.

**Comment by bjr70** [comment]
@monotasker do you need any more info on this or are we good to go? Do I need to go through and map the existing CORE types to the new types? Just let me know.
